subsys/pm/device_runtime.c

```diff
@@ -111,6 +111,16 @@ static int pm_device_request(const struct device *dev,
 		}
 	}
 
+	if (k_is_pre_kernel()) {
```

Comment:
I'm not convinced this is the correct way.
This means clock activation should be done on device side, on top of calling pm_device_get primitive (which at least, should be documented).
Instead, isn't it a way to run the worker routine code in a synchronous way at this stage ?

Reply:
This exactly code was part of my other pr
https://github.com/zephyrproject-rtos/zephyr/pull/33942/commits/64dac719dd5762dc46b5cb3e7d1774091feaf1ed

The reason for what I have not included it in this one is because this is a workaround from my understanding. At this point devices should have being initialized and powered on accordingly and this activation unnecessary. This is fixing an existing problem not in the proper way imho.

But as we don't have it fixed, I can cherry-pick this commit and once we have it done properly I remove this path.

Reply:
Btw, this is not about concurrence, but putting the device in a unexpected condition. From the device perspective is expected the init function be called before the power control function. The device may do some data initialization in the init that will be used in the power control  ....

Reply:
> At this point devices should have being initialized and powered on accordingly and this activation unnecessary.

No, this is not the case. you need to be able to use those drivers and use them means powering them up if they are not active. The issue was originally exactly that, we are trying to power on a device and cant be cause kernel is still not up, i.e. we try to deal with code designed for multithreds and concurrency that in this stage can be skipped.

Reply:
yep, but not in pre-kernel. Honestly I don't see why devices are not initialized in the right order at this point. Anyway, I've cherry-picked that commit

Reply:
what does initialized mean exactly? init() AFAIK does not power up a device necessarily..

Reply:
call init()